src/impl/mpispace/Kokkos_MPISpace_Ops.hpp

```diff
@@ -565,9 +577,9 @@ struct MPIDataElement<
   KOKKOS_INLINE_FUNCTION
   const_value_type operator+=(const_value_type &val) const {
     T tmp;
-    mpi_type_g(tmp, offset, pe, *win);
-    tmp += val;
-    mpi_type_p(tmp, offset, pe, *win);
+    // forward the pinky promise of not updating "val" of mpi.
+    mpi_type_atomic_add(const_cast<non_const_value_type &>(val), tmp, offset,
+                        pe, *win);
     return tmp;
```

Comment:
The value of `tmp` is undefined because there is no flush here. For single value operations MPI provides `MPI_Fetch_and_op` and it should be followed by `MPI_Win_flush_local`.

More generally, there is no way to hide the network latency for operators that return a result of an atomic op.

Also, you cannot guarantee that `val` will stay valid. If I write `a += 1` then `val` will almost certainly break the "pinky promise" and go out of scope/get changed.

And why only implement this for `operator+=` and not `operator*=`?

Reply:
Okay according to https://www.mpi-forum.org/docs/mpi-4.1/mpi41-report.pdf Page 577, line 47, `MPI_FETCH_AND_OP` is faster than `GET_ACCUMULATE`. I should use fetch_and_add where possible.